release/ray_release/cluster_manager/minimal.py

```diff
@@ -248,6 +251,39 @@ def create_cluster_compute(self, _repeat: bool = True):
                     f"ID {self.cluster_compute_id}"
                 )
 
+    @property
+    def cluster_compute_with_extra_tags(self) -> dict:
+        if self.extra_tags and self._get_cloud_provider() == "AWS":
+            cluster_compute = self.cluster_compute.copy()
+            if "aws" in cluster_compute:
+                cluster_compute["aws"] = deepcopy(cluster_compute["aws"])
+            else:
+                cluster_compute["aws"] = {}
+            cluster_compute["aws"].setdefault("TagSpecifications", [])
+            tag_specifications = cluster_compute["aws"]["TagSpecifications"]
+            for resource in self.extra_tags_resource_types:
+                resource_tags: dict = next(
+                    (
+                        x
+                        for x in tag_specifications
+                        if x.get("ResourceType", "") == resource
+                    ),
+                    None,
+                )
+                if resource_tags is None:
+                    resource_tags = {"ResourceType": resource, "Tags": []}
+                    tag_specifications.append(resource_tags)
+                tags = resource_tags["Tags"]
+                for key, value in self.extra_tags.items():
+                    tags.append({"Key": key, "Value": value})
```

Comment:
This is very hard to understand (tbh I couldn't be bothered to check the logic). Please rewrite this and add comments so someone looking at the code understands what is going on here.

It's ok to be not pythonic, I'd rather have explicit and easily understandable code than too many list comprehensions